clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java

```diff
@@ -52,51 +53,96 @@ public boolean hasCompatibleMagic(byte magic) {
         }
     }
 
-    private MemoryRecords convert(byte magic,
-                                  TimestampType timestampType,
-                                  CompressionType compressionType,
-                                  long logAppendTime,
-                                  List<LogRecord> records) {
-
-        LogRecord firstRecord = records.iterator().next();
-        long firstOffset = firstRecord.offset();
-        ByteBuffer buffer = ByteBuffer.allocate(estimatedSizeRecords(compressionType, records));
-        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, magic, compressionType, timestampType,
-                firstOffset, logAppendTime, 0L, (short) 0, 0);
-        for (LogRecord record : records)
-            builder.appendWithOffset(record.offset(), record.timestamp(), record.key(), record.value());
+    private MemoryRecords upconvertNoncompressedToV2AndAbove(byte magic, TimestampType upconvertTimestampType,
+                                                             List<LogEntryAndRecords> logEntryAndRecordsList) {
+        // when converting from older magic versions, we can pack the records in a message set into a single
+        // log entry. This gives us the benefit of the more efficient packing in v2 and above, and to avoid
+        // the increased overhead for singleton message sets.
+
+        List<LogRecord> logRecords = new ArrayList<>(logEntryAndRecordsList.size());
+        for (LogEntryAndRecords logEntryAndRecords : logEntryAndRecordsList)
+            logRecords.addAll(logEntryAndRecords.records);
+
+        LogEntryAndRecords firstEntryAndRecords = logEntryAndRecordsList.get(0);
+        LogEntry firstEntry = firstEntryAndRecords.entry;
+
+        final TimestampType timestampType;
+        if (firstEntry.magic() == LogEntry.MAGIC_VALUE_V0)
+            timestampType = upconvertTimestampType;
+        else
+            timestampType = firstEntry.timestampType();
+
+        long logAppendTime = timestampType == TimestampType.LOG_APPEND_TIME ? firstEntry.maxTimestamp() : LogEntry.NO_TIMESTAMP;
+
+        ByteBuffer buffer = ByteBuffer.allocate(EosLogEntry.sizeInBytes(firstEntryAndRecords.baseOffset, logRecords));
+        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, magic, firstEntry.compressionType(),
+                timestampType, firstEntryAndRecords.baseOffset, logAppendTime);
+
+        for (LogRecord logRecord : logRecords)
+            builder.append(logRecord);
         return builder.build();
     }
 
+    private ByteBuffer convertLogEntry(byte magic, ByteBuffer buffer, TimestampType upconvertTimestampType,
+                                       LogEntryAndRecords logEntryAndRecords) {
+        LogEntry entry = logEntryAndRecords.entry;
+        final TimestampType timestampType;
+        if (entry.magic() == LogEntry.MAGIC_VALUE_V0)
+            timestampType = upconvertTimestampType;
+        else
+            timestampType = entry.timestampType();
```

Comment:
What I had in mind with specific version -> version conversion is that, when we deprecate older formats we can simply remove the corresponding functions, also it makes the function easier to understand. Again, my main concern is not about correctness of duplicate code, but rather easy-readability of these functions since they were quite complex and mingled with multiple logic path and I'd like to see them more separated.

Will take another look at the modified function.